File: 数据结构与算法/栈与队列/stack.cs
/*
基于 C# 的栈实现
*/

using System.Diagnostics;

internal class Program
{
    private static void Main(string[] args)
    {
        // 做点测试
        Test<StackList<int>, int>([1, 5, 2, 4, 3]);
        Test<StackArray<int>, int>([1, 5, 2, 4, 3]);
        Console.WriteLine("All test passed.");
    }

    private static void Test<T, E>(IEnumerable<E> datas)
    where T : IStack<E>, new()
    where E : IEquatable<E>
    {
        Trace.Assert(datas.Count() >= 2);
        var stack = new T();
        stack.PushRange(datas);
        Trace.Assert(stack.Count() == datas.Count());
        Trace.Assert(stack.Top().Equals(datas.Last()));
        Trace.Assert(stack.Pop().Equals(datas.Last()));
        Trace.Assert(stack.Top().Equals(datas.ElementAt(datas.Count() - 2)));
        Trace.Assert(stack.Count() == datas.Count() - 1);
        stack.Clear();
        Trace.Assert(stack.Count() == 0);
        Trace.Assert(stack.IsEmpty());
        try
        {
            stack.Pop();
        }
        catch (Exception e)
        {
            Console.WriteLine($"Pop an empty stack and throw an exception: {e.Message}");
        }
        try
        {
            stack.Top();
        }
        catch (Exception e)
        {
            Console.WriteLine($"Get top element of an empty stack and throw an exception: {e.Message}");
        }
    }
}

// 栈接口
internal interface IStack<T>
{
    public void Push(T item);
    public T Pop();
    public T Top();
    public bool IsEmpty();
    public int Count();
    public void Clear();
    public void PushRange(IEnumerable<T> items);
}

// 基于链表的栈实现
internal class StackList<T> : IStack<T>
{
    private class Node
    {
        public required T data;
        public Node? next;
    }

    private Node? top; // 栈顶
    private int size; // 栈的大小

    public void Push(T item)
    {
        var node = new Node { data = item };
        if (top == null)
        {
            top = node;
        }
        else
        {
            node.next = top;
            top = node;
        }
        size++;
    }
    public T Pop()
    {
        if (top == null) throw new Exception("Stack is empty");
        T data = top.data;
        top = top.next;
        size--;
        return data;
    }
    public T Top()
    {
        if (top == null) throw new Exception("Stack is empty");
        return top.data;
    }
    public bool IsEmpty()
    {
        return size == 0;
    }
    public int Count()
    {
        return size;
    }
    public void Clear()
    {
        top = null;
        size = 0;
    }
    public void PushRange(IEnumerable<T> items)
    {
        items.ToList().ForEach(Push);
    }
}

// 基于数组的栈实现
internal class StackArray<T> : IStack<T>
{
    private readonly List<T> stack = []; // 栈

    public void Push(T item)
    {
        stack.Add(item);
    }
    public T Pop()
    {
        if (stack.Count == 0) throw new Exception("Stack is empty");
        T data = stack[^1];
        stack.RemoveAt(stack.Count - 1);
        return data;
    }
    public T Top()
    {
        if (stack.Count == 0) throw new Exception("Stack is empty");
        return stack[^1];
    }
    public bool IsEmpty()
    {
        return stack.Count == 0;
    }
    public int Count()
    {
        return stack.Count;
    }
    public void Clear()
    {
        stack.Clear();
    }
    public void PushRange(IEnumerable<T> items)
    {
        items.ToList().ForEach(Push);
    }
}

File: DataStructureAndAlgorithm/stack_go/stack.go
/*
基于 Golang 的栈实现
*/

package stackgo

// 栈接口
type Stack[T any] interface {
	Push(v *T)    // 入栈
	Pop() *T      // 出栈
	Top() *T      // 栈顶元素
	Size() uint64 // 栈大小
	Clear()       // 清空栈
}

// 基于切片的栈实现
type StackSlice[T any] struct {
	data []T
}

func (s *StackSlice[T]) Push(v *T) {
	if v == nil {
		return
	}
	s.data = append(s.data, *v)
}

func (s *StackSlice[T]) Pop() *T {
	if len(s.data) == 0 {
		return nil
	} else {
		v := s.data[len(s.data)-1]
		s.data = s.data[:len(s.data)-1]
		return &v
	}
}

func (s *StackSlice[T]) Top() *T {
	if len(s.data) == 0 {
		return nil
	} else {
		return &s.data[len(s.data)-1]
	}
}

func (s *StackSlice[T]) Size() uint64 {
	return uint64(len(s.data))
}

func (s *StackSlice[T]) Clear() {
	s.data = []T{}
}

// 链表节点
type Node[T any] struct {
	Val  T        // 节点值
	Next *Node[T] // 下一个节点
}

// 基于链表的栈实现
type StackLinkedList[T any] struct {
	TopNode *Node[T] // 栈顶节点
	Len     uint64   // 栈大小
}

func (s *StackLinkedList[T]) Push(v *T) {
	if v == nil {
		return
	}
	if s.TopNode == nil {
		s.TopNode = &Node[T]{Val: *v}
	} else {
		s.TopNode = &Node[T]{Val: *v, Next: s.TopNode}
	}
	s.Len++
}

func (s *StackLinkedList[T]) Pop() *T {
	if s.TopNode == nil {
		return nil
	} else {
		v := s.TopNode.Val
		s.TopNode = s.TopNode.Next
		s.Len--
		return &v
	}
}

func (s *StackLinkedList[T]) Top() *T {
	if s.TopNode == nil {
		return nil
	} else {
		return &s.TopNode.Val
	}
}

func (s *StackLinkedList[T]) Size() uint64 {
	return s.Len
}

func (s *StackLinkedList[T]) Clear() {
	s.TopNode = nil
	s.Len = 0
}
